### DISCUSSION TITLE
Does `?` type modifier generates ambiguous types?

### DISCUSSION THREAD

**Discussion by Vegetappen** [MichalLytek/typegraphql-prisma#264] [Q&A] [answered]
I'm playing around the basic example, Prisma schema `content String?` generates `content: string | null`(by prisma-client-js), but typegraphql-prisma generates `content?: string | null`. Also: 
type `PostCreateInput` generated by prisma-client-js is `content?: string | null`，
type `PostCreateInput` generated by typegraphql-prisma is `content?: string | undefined`, 
I think that optional parameter type `undefined` means set `NULL` in the database when creating an entity, but when I use types like `PostCreateInput`, the typescript compiler complain `Type 'undefined' is not assignable to type 'string | null'`. Why these types looks different and how to deal with the `undefined` when create or update something in a resolver? Really need advice.

**Comment by MichalLytek** [comment, marked as answer]
Do you refer to input or output types?

In GraphQL (JS), when you don't provide argument or field for nullable input, its value in resolvers is `undefined`. When you explicitly provide null, the value will be null.

For output types, when you provide `undefined` for the nullable field value, it will be returned as `null` in JSON response.

**Reply by Vegetappen** [reply]
Thanks for replying. I don't understand why does the input type not accept `null` literal?
```typescript
content?: string | undefined
```

**Reply by MichalLytek** [reply]
Fair point. GraphQL input field can be null and that null can hurt Prisma 🤔

**Reply by Vegetappen** [reply]
I don't know how others code, I would explicitly set some field to `null` instead of something like `''`(an empty string). I think it is serious and package does not generate the types I want.

**Reply by MichalLytek** [reply]
You can send `null` via GraphQL, it's just not reflected in the generated TS types but it works in runtime.

For create/update Prisma allows null but for where and other optional args - does not.

**Reply by Vegetappen** [reply]
> it's just not reflected in the generated TS types but it works in runtime

https://github.com/MichalLytek/typegraphql-prisma/blob/b589395175b857ec0f8d120511f3a9b4ba97167c/Readme.md?plain=1#L5

I'm going to call it an end. Thanks for replying and all your contribution.

**Reply by MichalLytek** [reply]
> I'm going to call it an end.

Are you looking for a solution to your issue or just wanted to complain and offend to make yourself better?

**Reply by Vegetappen** [reply]
My issue? It was just basic types an field which accepts string or null. Complain? Yes. It works incorrectly from my view. If offend, my apologize. Feeling better? No, just have a try on the basic example.

**Reply by MichalLytek** [reply]
As I said, it's not so easy. For Prisma and optional properties, sometimes null is acceptable (like a column value) but sometimes not (where conditions). On the other hand, GraphQL is allowing null, thus giving the conflicts.

So AFAIR two years ago I took the shortcut by ignoring null in type definition, so that the types are usable in custom resolvers with Prisma Client. If they would be make 100% correct, there will be tons of red underlines because of invalid null, forcing everyone to manually map the input values to undefined on null.

Neither `graphql-js` or TypeGraphQL support transforming input types fields, so while I could handle nulls in generated crud resolvers, that would not work for custom resolvers.

So yes, it works incorrectly even in basic example. So far nobody else complained about that, so I don't find that a critical issue.

**Reply by Vegetappen** [reply]
As reference, below types were generated by Prisma: 

```typescript
export type PostWhereInput = {
  AND?: Enumerable<PostWhereInput>;
  OR?: Enumerable<PostWhereInput>;
  NOT?: Enumerable<PostWhereInput>;
  id?: IntFilter | number;
  createdAt?: DateTimeFilter | Date | string;
  updatedAt?: DateTimeFilter | Date | string;
  title?: StringFilter | string;
  content?: StringNullableFilter | string | null;
};
```

```typescript
export type StringNullableFilter = {
  equals?: string | null;
  in?: Enumerable<string> | null;
  notIn?: Enumerable<string> | null;
  lt?: string;
  lte?: string;
  gt?: string;
  gte?: string;
  contains?: string;
  startsWith?: string;
  endsWith?: string;
  mode?: QueryMode;
  not?: NestedStringNullableFilter | string | null;
};
```

It looks good from an ORM view and there is [documented use case](https://www.prisma.io/docs/concepts/components/prisma-client/null-and-undefined#use-case-null-and-undefined-in-a-graphql-resolver) that skip update an field which not accepts `null`. It transform(transforming input fields as you said) `null` to `undefined` then put `undefined` to do nothing. So you took the shortcut for 
generated crud resolvers and also for custom resolvers. If you add validation(which should be implemented by developers), you will not need that shortcut, right? Do I kind of catch what you mean?
